[FancyZones] Update app zone history on window removal

vldmr11080 opened the pull request:
<!-- Enter a brief description/summary of your PR here. What does it fix/what does it change/how was it tested (even manually, if necessary)? -->
## Summary of the Pull Request
If there are several windows of same application zoned in same work area, app zone history is remembered for only one of them. If that window is removed from zone, we need to try to update app zone history with zone set from some other window (of same application). If there are several options, windows are sorted by creation time and oldest (earliest started process).

<!-- Please review the items on the PR checklist before submitting-->
## PR Checklist
* [x] Applies to #4356
* [x] CLA signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/PowerToys) and sign the CLA

<!-- Describe how you validated the behavior. Add automated tests wherever possible, but list manual validation steps taken as well -->
## Validation Steps Performed
1. Create zone layout on one of the monitors (it behaves same for both primary and secondary monitors). For example three zone column.
2. Assign window to zone 0.
3. Open on same monitor another window of same application (gets zoned on opening to zone 0).
4. Zone that new window to a zone 1. **History still shows zone 0 -> expected**
5. Move first window out of zone. **History updated, now history shows zone 1 (belonging to remaining instance) -> expected**
6. Now zone first window to zone 2. **History still shows zone 1 -> expected**
7. Move second window out of zone. **History updated, now history shows zone 2 (belonging to remaining instance) -> expected**

Essentially, when window is removed from zone, if that zone was saved in app zone history, we try to find another window of same application, within same work are to update history. If there are more than one, we use zone set from "oldest" window (earliest started process).

Try same procedure with more than two instances of same application to confirm behavior when oldest window zone is selected for new zone in history.

ivan100sic commented on src/modules/fancyzones/lib/JsonHelpers.cpp:
```diff
@@ -32,6 +34,33 @@ namespace
+    bool GetProcessCreationTime(DWORD processId, FILETIME* creationTime)
+    {
+        bool result{ false };
+        HANDLE hProcess = OpenProcess(PROCESS_QUERY_INFORMATION | PROCESS_VM_READ, FALSE, processId);
```
I don't think we need `PROCESS_VM_READ`

vldmr11080 replied:
Removed!

ivan100sic commented on src/modules/fancyzones/lib/JsonHelpers.cpp:
```diff
@@ -442,6 +471,72 @@ namespace JSONHelpers
+                size_t placedWindowZoneStamp = reinterpret_cast<size_t>(::GetProp(placedWindow.second, MULTI_ZONE_STAMP));
+                if (placedWindowZoneStamp == windowZoneStamp)
+                {
+                    // if there is another instance of same appliaction placed in the same zone don't erase history
```
Typo: appliaction

vldmr11080 replied:
Fixed.

vldmr11080 commented:
> Nice work, I tested manually as described and checked settings file. All works as expected. Unit tests passed.
> However in my opinion `HandleWindowRemovalFromZone` name is unclear. Also method himself makes several things: checks if process should be removed from app zone history map and updates `AppZoneHistoryData`. Could you please refactor this method and split into smaller parts?

I've split it into three smaller more meaningful methods. Please take a look again.

vldmr11080 commented:
> LGTM, I checked the validation scenario and unit tests.
> 
> I'll take another look if you decide to refactor anything.

Thank You. Let me just rebase on latest master branch (and resolve this conflict).

SeraphimaZ commented:
Code looks better, thanks. 
I tested with notepad and explorer and before rebasing and refactoring both worked same. Now with notepad it works as expected, but with explorer settings are updated every time when any app instance is assigned to a zone.

1. Create zone layout on one of the monitors (it behaves same for both primary and secondary monitors). For example three zone column.
2. Assign window to zone 0.
3. Open on same monitor another window of same application (gets zoned on opening to zone 0).
4. Zone that new window to a zone 1. _History still shows zone 0 -> expected._ **History changed to 1 -> actual.**
5. Move first window out of zone. _History updated, now history shows zone 1 (belonging to remaining instance) -> expected._
6. Now zone first window to zone 2. _History still shows zone 1 -> expected._ **History changed to 2 -> actual.**
7. Move second window out of zone. _History updated, now history shows zone 2 (belonging to remaining instance) -> expected._

vldmr11080 commented:
> Code looks better, thanks.
> I tested with notepad and explorer and before rebasing and refactoring both worked same. Now with notepad it works as expected, but with explorer settings are updated every time when any app instance is assigned to a zone.
> 
> 1. Create zone layout on one of the monitors (it behaves same for both primary and secondary monitors). For example three zone column.
> 2. Assign window to zone 0.
> 3. Open on same monitor another window of same application (gets zoned on opening to zone 0).
> 4. Zone that new window to a zone 1. _History still shows zone 0 -> expected._ **History changed to 1 -> actual.**
> 5. Move first window out of zone. _History updated, now history shows zone 1 (belonging to remaining instance) -> expected._
> 6. Now zone first window to zone 2. _History still shows zone 1 -> expected._ **History changed to 2 -> actual.**
> 7. Move second window out of zone. _History updated, now history shows zone 2 (belonging to remaining instance) -> expected._

Hm, that is true, and I see why do we have that scenario. All explorer instances (windows) are essentially bound to same process id, unlikely to other applications, where every new window makes process of his own. I'll find approach to handle this in more robust manner.

vldmr11080 commented:
> > Code looks better, thanks.
> > I tested with notepad and explorer and before rebasing and refactoring both worked same. Now with notepad it works as expected, but with explorer settings are updated every time when any app instance is assigned to a zone.
> > 
> > 1. Create zone layout on one of the monitors (it behaves same for both primary and secondary monitors). For example three zone column.
> > 2. Assign window to zone 0.
> > 3. Open on same monitor another window of same application (gets zoned on opening to zone 0).
> > 4. Zone that new window to a zone 1. _History still shows zone 0 -> expected._ **History changed to 1 -> actual.**
> > 5. Move first window out of zone. _History updated, now history shows zone 1 (belonging to remaining instance) -> expected._
> > 6. Now zone first window to zone 2. _History still shows zone 1 -> expected._ **History changed to 2 -> actual.**
> > 7. Move second window out of zone. _History updated, now history shows zone 2 (belonging to remaining instance) -> expected._
> 
> Hm, that is true, and I see why do we have that scenario. All explorer instances (windows) are essentially bound to same process id, unlikely to other applications, where every new window makes process of his own. I'll find approach to handle this in more robust manner.

@SeraphimaZ please give it another try. Call to `UpdateProcessIdToHandleMap` should always be called after check to `IsAnotherWindowOfApplicationInstanceZoned` and that was lacking in previous implementation. I've also confirmed that it don't breaks functionality for other applications (give it a check also).

SeraphimaZ commented:
I played with it and not sure if it scenario worked right. Should history be updated when the first window is moved to another zone?

Notepad:
1. Assign window to zone 0.
2. Open on same monitor another window of same application.
3. Zone that new window to a zone 1. _History still shows zone 0 -> expected._
4. Move first window to zone 2. _History shows zone 2 -> expected?_ **History not updated -> actual.**

Explorer:
1. Assign window to zone 0.
2. Open on same monitor another window of same application.
3. Zone that new window to a zone 1. _History still shows zone 0 -> expected._
4. Move first window to zone 2. _History shows zone 2 -> expected?_ **History shows zone 2-> actual.**

And regarding previous problem with explorer instances:
1. Assign window to zone 0.
2. Open on same monitor another window of same application.
3. Zone that new window to a zone 1. _History still shows zone 0 -> expected._
4. Move first window out of zone. _History shows zone 1 -> expected._ **History erased -> actual**
5. Now zone first window to zone 2. _History still shows zone 1 -> expected._ **History changed to 2 -> actual.**
6. Move second window out of zone. _History shows zone 2 -> expected._

enricogior commented:
```
1. Assign window to zone 0.
2. Open on same monitor another window of same application.
3. Zone that new window to a zone 1. History still shows zone 0 -> expected.
4. Move first window to zone 2. History shows zone 2 -> expected? History not updated -> actual.
```
I agree, this is unexpected, I would suggest that we always set the current zones has last zone:

1. Assign windows to zone 0: _history is created to point to zone 0_
2. Open on same monitor another window of same application.
3. Zone that new window to zone 1: _history is updated to point to zone 1_
4. Move first window to zone 2: _history is updated to point to zone 2_

This is the simplest logic and users will easily understand.

enricogior commented:
Two monitors: on monitor one the window is zoned, on monitor two same app is zoned and then moved out of the zone and still on monitor two.
The expected behavior is that the history for monitor two is removed, instead this happens:

![image](https://user-images.githubusercontent.com/3206696/86728146-74451580-c02c-11ea-87c6-d542a78b0154.png)

vldmr11080 commented:
> ```
> 1. Assign window to zone 0.
> 2. Open on same monitor another window of same application.
> 3. Zone that new window to a zone 1. History still shows zone 0 -> expected.
> 4. Move first window to zone 2. History shows zone 2 -> expected? History not updated -> actual.
> ```
> 
> I agree, this is unexpected, I would suggest that we always set the current zones has last zone:
> 
> 1. Assign windows to zone 0: _history is created to point to zone 0_
> 2. Open on same monitor another window of same application.
> 3. Zone that new window to zone 1: _history is updated to point to zone 1_
> 4. Move first window to zone 2: _history is updated to point to zone 2_
> 
> This is the simplest logic and users will easily understand.

I wanted to suggest that, since this is quite complicated. And one more addition, I'll change my logic in `HandleWindowRemovalFromZone` so that I don't look for earlies started process to update zone history, but rather for latest.

One more thing, I don't think it will be possible (without breaking existing functionality) to make for example cmd.exe and explorer.exe to behave the same. That is just not possible. Multiple windows of explorer.exe are bound to same process, while every window of cmd.exe is separate process (this is how most apps behave).

enricogior commented:
@vldmr11080 

>  I don't look for earlies started process to update zone history, but rather for latest.

Yes, this is what I meant as well.

vldmr11080 commented:
Removed all complex logic that handles window removal from zone. Just simple continuous updates to zone history whenever new window is zoned. @enricogior @SeraphimaZ PTAL.
